native/src/main/java/io/ballerina/stdlib/io/nativeimpl/CsvChannelUtils.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2019 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.ballerina.stdlib.io.nativeimpl;
+
+import io.ballerina.runtime.api.TypeTags;
+import io.ballerina.runtime.api.creators.TypeCreator;
+import io.ballerina.runtime.api.creators.ValueCreator;
+import io.ballerina.runtime.api.types.Field;
+import io.ballerina.runtime.api.types.StructureType;
+import io.ballerina.runtime.api.types.Type;
+import io.ballerina.runtime.api.types.UnionType;
+import io.ballerina.runtime.api.utils.StringUtils;
+import io.ballerina.runtime.api.values.BObject;
+import io.ballerina.runtime.api.values.BStream;
+import io.ballerina.runtime.api.values.BString;
+import io.ballerina.runtime.api.values.BTypedesc;
+import io.ballerina.stdlib.io.utils.IOUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static io.ballerina.stdlib.io.nativeimpl.RecordChannelUtils.getAllRecords;
+import static io.ballerina.stdlib.io.nativeimpl.RecordChannelUtils.hasNext;
+import static io.ballerina.stdlib.io.utils.IOConstants.CSV_RETURN_TYPE;
+import static io.ballerina.stdlib.io.utils.IOUtils.getIOPackage;
+
+
+/**
+ * This class hold Java inter-ops bridging functions for io# *CSVChannel/*RTextRecordChannel.
+ *
+ */
+
+public class CsvChannelUtils {
+    private static final Logger log = LoggerFactory.getLogger(CsvChannelUtils.class);
+    
+    public static Object fileReadCsv(BString path, int skipHeaders, BTypedesc typeDesc) {
+        BObject byteChannel = (BObject) ByteChannelUtils.openReadableFile(path);
+
+        BObject characterChannel = ValueCreator.createObjectValue(getIOPackage(), "ReadableByteChannel");
+        CharacterChannelUtils.initCharacterChannel(characterChannel, byteChannel, StringUtils.fromString("UTF-8"));
+
+        BString fs = StringUtils.fromString(",");
+        BString rs = StringUtils.fromString("");
+        BString format = StringUtils.fromString("CSV");
+        BObject textRecordChannel = ValueCreator.createObjectValue(getIOPackage(), "ReadableByteChannel");
+        RecordChannelUtils.initRecordChannel(textRecordChannel, characterChannel, fs, rs, format);
+
+        textRecordChannel.addNativeData(CSV_RETURN_TYPE, typeDesc);
```

Comment:
The refernce type check should be done before this. Fix in other places as well

Reply:
Can you kindly explain more? I'm not sure what you meant by type check

Reply:
Lets have a chat over zoom on this. It is complex to explain with comments.